File: leetcode/medium/15-3Sum.py
def threeSum(nums):
  nums.sort()
  result = []

  for index, num in enumerate(nums):
    if index > 0 and num == nums[index - 1]: continue
    left = index + 1
    right = len(nums) - 1

    while left < right:
      equalsZero = num + nums[left] + nums[right]

      if equalsZero == 0:
        result.append([num, nums[left], nums[right]])
        left += 1
        while nums[left] == nums[left - 1] and left < right:
          left += 1
      elif equalsZero < 0:
        left += 1
      else:
        right -= 1

  return result

File: leetcode/medium/15-3Sum.js
// Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
// Notice that the solution set must not contain duplicate triplets.

// Input: nums = [-1,0,1,2,-1,-4]
// Output: [[-1,-1,2],[-1,0,1]]
// Explanation:
// nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
// nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
// nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
// The distinct triplets are [-1,0,1] and [-1,-1,2].
// Notice that the order of the output and the order of the triplets does not matter.

// Time: O(n^2)
// Space: O(n)

const threeSum = (nums) => {
  nums.sort((a, b) => a - b);
  let result = [];

  for (let i = 0; i < nums.length; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) continue;

    let left = i + 1;
    let right = nums.length - 1;

    while (left < right) {
      let equalsZero = nums[i] + nums[left] + nums[right];

      if (equalsZero === 0) {
        result.push([nums[i], nums[left], nums[right]]);
        left++;
        while (nums[left] === nums[left - 1] && left < right) {
          left++;
        }
      } else if (equalsZero < 0) {
        left++;
      } else {
        right--;
      }
    }
  }
  return result;
};